packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/useConnectionCandidatesQuery.ts

```diff
@@ -0,0 +1,116 @@
+          description {
+            ... on DiagramDescription {
+              palette(diagramElementId: $diagramElementId) {
+                tools {
+                  __typename
+                  ... on SingleClickOnTwoDiagramElementsTool {
+                    candidates {
+                      sources {
+                        id
+                      }
+                      targets {
+                        id
+                      }
+                    }
+                  }
+                }
+                toolSections {
+                  tools {
+                    __typename
+                    ... on SingleClickOnTwoDiagramElementsTool {
+                      candidates {
+                        sources {
+                          id
+                        }
+                        targets {
+                          id
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+`;
+
+const isDiagramDescription = (
+  representationDescription: GQLRepresentationDescription
+): representationDescription is GQLDiagramDescription => representationDescription.__typename === 'DiagramDescription';
+const isSingleClickOnTwoDiagramElementsTool = (tool: GQLTool): tool is GQLSingleClickOnTwoDiagramElementsTool =>
+  tool.__typename === 'SingleClickOnTwoDiagramElementsTool';
+
+export const useConnectionCandidatesQuery = (
+  editingContextId: string,
+  diagramId: string,
+  nodeId: string
+): GQLNodeDescription[] | null => {
+  const [candidates, setCandidates] = useState<GQLNodeDescription[] | null>(null); // null if not yet loaded
+  const { addErrorMessage } = useMultiToast();
+
+  const { loading, data, error } = useQuery<GQLGetToolSectionsData, GQLGetToolSectionsVariables>(getToolSectionsQuery, {
+    variables: {
+      editingContextId,
+      diagramId,
+      diagramElementId: nodeId,
+    },
+  });
+
+  useEffect(() => {
+    if (!loading && data) {
+      const diagramDescription: GQLRepresentationDescription = data.viewer.editingContext.representation.description;
+      const nodesCandidates: GQLNodeDescription[] = [];
+      if (isDiagramDescription(diagramDescription)) {
+        diagramDescription.palette.tools.filter(isSingleClickOnTwoDiagramElementsTool).forEach((tool) => {
+          tool.candidates.forEach((candidate) => nodesCandidates.push(...candidate.targets));
+        });
+        diagramDescription.palette.toolSections.forEach((toolSection) => {
+          toolSection.tools.filter(isSingleClickOnTwoDiagramElementsTool).forEach((tool) => {
+            tool.candidates.forEach((candidate) => nodesCandidates.push(...candidate.targets));
+          });
+        });
+      }
+      setCandidates(nodesCandidates);
+
+      if (error) {
+        addErrorMessage(error.message);
+      }
+    }
+  }, [loading, data, error]);
```

Comment:
Unless I'm mistaken, here you have the same small issue, everything here could be derived without any state of useEffect (you would need one for the error message only):


```
      const diagramDescription: GQLRepresentationDescription | null = data?.viewer.editingContext.representation.description ?? null;
      const candidates: GQLNodeDescription[] = [];
      if (isDiagramDescription(diagramDescription)) {
        diagramDescription.palette.tools.filter(isSingleClickOnTwoDiagramElementsTool).forEach((tool) => {
          tool.candidates.forEach((candidate) => nodesCandidates.push(...candidate.targets));
        });
        diagramDescription.palette.toolSections.forEach((toolSection) => {
          toolSection.tools.filter(isSingleClickOnTwoDiagramElementsTool).forEach((tool) => {
            tool.candidates.forEach((candidate) => nodesCandidates.push(...candidate.targets));
          });
        });
      }

      useEffect(() => {
        if (error) {
          addErrorMessage(error.message);
        }
      }, [error]);
```

We have been doing this pretty much everywhere :)